File: churn-demo/new/ledger.py
class Ledger:
    def __init__(self, name):
        self.name = name
        self.previous_name = ""
        self.balance = 0.0
        self.limit = 0.0
        self.interest_paid = 0.0
        self.transfers = 0
        self.pending_holds = 0
        self.cleared_items = 0
        self.flagged_items = 0
        self.disputed_items = 0
        self.entries = []
        self.summary = {}

    def post(self, status, amount):
        self.entries.append((status, amount))
        self.balance = self.balance + amount
        self.log("posted")

    def credit(self, amount):
        self.post("posted", amount)
        self.cleared_items += 1

    def debit(self, amount):
        self.post("posted", -amount)
        self.cleared_items += 1

    def hold(self, amount):
        self.post("pending", amount)
        self.pending_holds += 1

    def dispute(self, amount):
        self.post("disputed", amount)
        self.disputed_items += 1
        self.flagged_items += 1

    def transfer(self, to, amount):
        self.debit(amount)
        to.credit(amount)
        self.transfers += 1

    def apply_interest(self, rate):
        interest = self.balance * rate
        self.post("posted", interest)
        self.interest_paid += interest

    def flag_large(self, threshold):
        n = 0
        for status, amount in self.entries:
            if abs(amount) > threshold:
                n += 1
        self.flagged_items += n
        return n

    def largest_entry(self):
        best = 0.0
        for status, amount in self.entries:
            if abs(amount) > abs(best):
                best = amount
        return best

    def statement_line(self, i):
        status, amount = self.entries[i]
        return "%s:%.2f" % (status, amount)

    def rename(self, new_name):
        self.previous_name = self.name
        self.name = new_name
        self.log("renamed")

    def net_flow(self):
        return self.sum_signed(+1) - self.sum_signed(-1)

    def reset(self):
        self.entries = []
        self.summary = {}
        self.balance = 0.0
        self.log("reset")

    def sum_signed(self, sign):
        total = 0.0
        for status, amount in self.entries:
            if (amount >= 0) == (sign > 0):
                total += abs(amount)
        return total

    def count_status(self, wanted):
        n = 0
        for status, amount in self.entries:
            if status == wanted:
                n += 1
        return n

    def set_limit(self, limit):
        self.limit = limit
        self.log("limit changed")

    def over_limit(self):
        return self.balance < -self.limit

    def entry_count(self):
        return len(self.entries)

    def summary_value(self, key):
        return self.summary.get(key, 0.0)

    def reconcile(self):
        opening = self.balance
        credits = self.sum_signed(+1)
        debits = self.sum_signed(-1)
        closing = opening + credits - debits
        drift = closing - self.balance
        posted = self.count_status("posted")
        pending = self.count_status("pending")
        disputed = self.count_status("disputed")
        total = posted + pending + disputed
        average = (credits + debits) / total if total > 0 else 0.0
        self.summary["opening"] = opening
        self.summary["credits"] = credits
        self.summary["debits"] = debits
        self.summary["closing"] = closing
        self.summary["drift"] = drift
        self.summary["posted"] = posted
        self.summary["pending"] = pending
        self.summary["disputed"] = disputed
        self.summary["average"] = average
        self.balance = closing

    def log(self, m):
        pass

File: churn-demo/new/Ledger.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Ledger {
    private String name;
    private String previousName;
    private double balance;
    private double limit;
    private double interestPaid = 0.0;
    private int transfers = 0;
    private int pendingHolds = 0;
    private int clearedItems = 0;
    private int flaggedItems = 0;
    private int disputedItems = 0;
    private final List<Entry> entries = new ArrayList<>();
    private final Map<String, Double> summary = new HashMap<>();

    static final class Entry {
        final String status;
        final double amount;
        Entry(String status, double amount) { this.status = status; this.amount = amount; }
    }

    public Ledger(String name) {
        this.name = name;
        this.balance = 0.0;
        this.limit = 0.0;
    }

    public void post(String status, double amount) {
        entries.add(new Entry(status, amount));
        balance = balance + amount;
        log("posted");
    }

    public void credit(double amount) {
        post("posted", amount);
        clearedItems += 1;
    }

    public void debit(double amount) {
        post("posted", -amount);
        clearedItems += 1;
    }

    public void hold(double amount) {
        post("pending", amount);
        pendingHolds += 1;
    }

    public void dispute(double amount) {
        post("disputed", amount);
        disputedItems += 1;
        flaggedItems += 1;
    }

    public void transfer(Ledger to, double amount) {
        debit(amount);
        to.credit(amount);
        transfers += 1;
    }

    public void applyInterest(double rate) {
        double interest = balance * rate;
        post("posted", interest);
        interestPaid += interest;
    }

    public int flagLarge(double threshold) {
        int n = 0;
        for (Entry e : entries) {
            if (Math.abs(e.amount) > threshold) n += 1;
        }
        flaggedItems += n;
        return n;
    }

    public double largestEntry() {
        double best = 0.0;
        for (Entry e : entries) {
            if (Math.abs(e.amount) > Math.abs(best)) best = e.amount;
        }
        return best;
    }

    public String statementLine(int i) {
        Entry e = entries.get(i);
        return e.status + ":" + e.amount;
    }

    public void rename(String newName) {
        previousName = name;
        name = newName;
        log("renamed");
    }

    public double netFlow() {
        return sumSigned(+1) - sumSigned(-1);
    }

    public void reset() {
        entries.clear();
        summary.clear();
        balance = 0.0;
        log("reset");
    }

    public double sumSigned(int sign) {
        double sum = 0.0;
        for (Entry e : entries) {
            if ((e.amount >= 0) == (sign > 0)) sum += Math.abs(e.amount);
        }
        return sum;
    }

    public int countStatus(String status) {
        int n = 0;
        for (Entry e : entries) {
            if (e.status.equals(status)) n += 1;
        }
        return n;
    }

    public void setLimit(double limit) {
        this.limit = limit;
        log("limit changed");
    }

    public boolean overLimit() {
        return balance < -limit;
    }

    public String getName() {
        return name;
    }

    public double getBalance() {
        return balance;
    }

    public int entryCount() {
        return entries.size();
    }

    public double summaryValue(String key) {
        return summary.getOrDefault(key, 0.0);
    }

    public void reconcile() {
        double opening = balance;
        double credits = sumSigned(+1);
        double debits = sumSigned(-1);
        double closing = opening + credits - debits;
        double drift = closing - balance;
        int postedCount = countStatus("posted");
        int pendingCount = countStatus("pending");
        int disputedCount = countStatus("disputed");
        int total = postedCount + pendingCount + disputedCount;
        double average = total > 0 ? (credits + debits) / total : 0.0;
        summary.put("opening", opening);
        summary.put("credits", credits);
        summary.put("debits", debits);
        summary.put("closing", closing);
        summary.put("drift", drift);
        summary.put("posted", (double) postedCount);
        summary.put("pending", (double) pendingCount);
        summary.put("disputed", (double) disputedCount);
        summary.put("average", average);
        balance = closing + interestPaid;
    }

    public void resetCounters() { pendingHolds = 0; clearedItems = 0; flaggedItems = 0; disputedItems = 0; }

    private void log(String m) {}
}
